### PR TITLE
TLUNIZIN-910 Add Link header to response where applicable

### PR DESCRIPTION
**Pull request by dovek**
Due: 3/23
@gsilver 
@dlhaines 
See JIRA for regular expression change.

### CODE REVIEW

File: src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java [hunk 1]
```diff
@@ -914,7 +953,13 @@ private boolean isCrosslistAllowed(HttpServletRequest request,
 
 		clientRequest = new HttpGet(crosslistApiCall);
 
-		BufferedReader rd = processApiCall(clientRequest);
+		HttpResponse canvasResponse = processApiCall(clientRequest);
+		BufferedReader rd = null;
+		try {
+			rd = new BufferedReader(new InputStreamReader(canvasResponse.getEntity().getContent()));
```
Comment (dlhaines) [inline src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java]: What happens if getEntity is null?

File: src/main/webapp/assets-lti/js/factories/factories.js [hunk 2]
```diff
@@ -118,11 +118,13 @@ canvasSupportApp.factory('Sections', function ($http) {
     getSectionsForCourseId: function (courseId, context) {
     	if ( context === true){
         //REGEXINFO: canvas.api.getallsections.per.course.regex
-    		var url = '/canvasCourseManager/manager/api/v1/courses/course_id/sections?per_page=100&_='+ generateCurrentTimestamp();
+    	//modified per_page query for testing
+    		var url = '/canvasCourseManager/manager/api/v1/courses/course_id/sections?per_page=2&_='+ generateCurrentTimestamp();
```
Comment (dlhaines) [inline src/main/webapp/assets-lti/js/factories/factories.js]: Shouldn't check in version modified for testing.  Should make the per_page value a property that we default to 100.

File: src/main/webapp/assets-lti/js/factories/factories.js [hunk 3]
```diff
@@ -118,11 +118,13 @@ canvasSupportApp.factory('Sections', function ($http) {
...
     	} 
     	else{
         //REGEXINFO: canvas.api.getallsections.per.course.regex
-    		var url = '/canvasCourseManager/manager/api/v1/courses/' + courseId + '/sections?per_page=100&_='+ generateCurrentTimestamp();	
+    	//modified per_page query for testing
+    		var url = '/canvasCourseManager/manager/api/v1/courses/' + courseId + '/sections?per_page=2&_='+ generateCurrentTimestamp();
```
Comment (dlhaines) [inline src/main/webapp/assets-lti/js/factories/factories.js]: See above.

File: src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java [hunk 4]
```diff
@@ -630,7 +635,27 @@ private void getCanvasResponse(HttpServletRequest request,
...
+		BufferedReader rd = new BufferedReader(new InputStreamReader(canvasResponse.getEntity().getContent()));
+
+		String linkValue = null;
+		
+		Header[] headers = canvasResponse.getAllHeaders();
+		for (Header header : headers) {
+			M_log.debug("Key : " + header.getName() 
+			      + " ,Value : " + header.getValue());
+			if(header.getName().equals("Link")){
+				linkValue = header.getValue();
```
Comment (dlhaines) [inline src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java]: Can you break here?  That makes it clear that you are only looking for 1 values.  

If you need to get another specific header, consider making this a utility method.

File: src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java [hunk 5]
```diff
@@ -729,6 +754,13 @@ public void displaySessionAttributes(HttpServletRequest request) {
 		}
 	}
 
+	public void displayRequestHeaders(HttpServletRequest request) {
+		Enumeration headerNames = request.getHeaderNames();
```
Comment (dlhaines) [inline src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java]: Check if debug is enabled as the first line of method so don't do anything if not debugging.  Also makes it clear that this method is only for debugging.

File: src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java [hunk 6]
```diff
@@ -729,6 +754,13 @@ public void displaySessionAttributes(HttpServletRequest request) {
...
+		while(headerNames.hasMoreElements()){
+			M_log.debug("HEADER NAME: " + headerNames.nextElement());
+		}
+	}
+
 	public void testMpathwaysCall(PrintWriter out) {
```
Comment (dlhaines) [inline src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java]: Should testMpathwasyCall be in production code?

Comment (dovek) [inline src/main/java/edu/umich/ctools/sectionsUtilityTool/SectionsUtilityToolServlet.java]: Yes, this was used in stubbing.

**Pull request merged**